### PR TITLE
Invoking ImgDetails Post APi in buildPush DockerV2 task

### PR DESCRIPTION
**Pull request by dabasnidhi**
When we run the DockerV2 task to build and push an image to a container registry, we will call the API -  [AddImageDetailsAsync](https://dev.azure.com/mseng/AzureDevOps/_git/AzureDevOps?path=%2FDeployment%2FClient%2FWebApi%2FGenerated%2FDeploymentHttpClient.cs&version=GBmaster&line=70&lineStyle=plain&lineEnd=70&lineStartColumn=27&lineEndColumn=47) to post the [image details as per Grafeas format ](https://dev.azure.com/mseng/AzureDevOps/_git/AzureDevOps?path=%2FDeployment%2FClient%2FWebApi%2FContracts%2FImageDetails.cs&version=GBmaster) to Deployment DB (tables Deployment.tbl_Note and Deployment.tbl_Occurrence).
This data will then be used to show the image details in the UI via [data providers](https://dev.azure.com/mseng/AzureDevOps/_git/AzureDevOps?path=%2FDeployment%2FService%2FPlugins%2FDataProviders&version=GBmaster)

### CODE REVIEW

File: Tasks/DockerV2/dockerpush.ts [hunk 1]
```diff
@@ -79,3 +99,58 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
 
     return promise;
 }
+
+async function publishToImageMetadataStore(connection: ContainerConnection, imageName: string, tags: string[], digest: string, dockerFilePath: string): Promise<any> {
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: Move this method to pipelineutils.ts. https://github.com/Microsoft/azure-pipelines-tasks/blob/master/Tasks/Common/docker-common/pipelineutils.ts

File: Tasks/DockerV2/dockerpush.ts [hunk 2]
```diff
@@ -79,3 +99,58 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
...
+            "layerInfo": layers,
+            "buildId": buildId,
+            "buildVersion": buildVersion,
+            "buildDefinitionName": buildDefinitionName,
+            "buildDefinitionId": buildDefinitionId
+        }
+    );
+
+    const request = new WebRequest();
+    request.uri = requestUrl;
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: Dont you need the accessToken here?

File: Tasks/DockerV2/dockerpush.ts [hunk 3]
```diff
@@ -79,3 +99,58 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
...
+    const dockerFileContent: string = fs.readFileSync(dockerFilePath, 'utf-8').toString();
+    const baseImageName = getBaseImageName(dockerFileContent);
+    const layers = dockerCommandUtils.getLayers(connection, imageUri);
+    const buildId = parseInt(tl.getVariable("Build.BuildId"));
+    const buildDefinitionName = tl.getVariable("Build.DefinitionName");
+    const buildVersion = tl.getVariable("Build.BuildNumber");
+    const buildDefinitionId = tl.getVariable("System.DefinitionId");
+
+    const requestUrl: string = tl.getVariable("System.TeamFoundationCollectionUri") + "/" + tl.getVariable("System.TeamProject") + "/_apis/deployment/imagedetails?api-version=5.0-preview.1";
+    const requestBody: string = JSON.stringify(
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: Move code below this line to a separate method.

File: Tasks/DockerV2/dockerpush.ts [hunk 4]
```diff
@@ -79,3 +99,58 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
 
     return promise;
 }
+
+async function publishToImageMetadataStore(connection: ContainerConnection, imageName: string, tags: string[], digest: string, dockerFilePath: string): Promise<any> {
+    const imageUri: string = getResourceName(imageName, digest);
+    const dockerFileContent: string = fs.readFileSync(dockerFilePath, 'utf-8').toString();
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: We should read the docker file in getBaseImageName method itself. Maybe have another overload of it.

File: Tasks/DockerV2/dockerpush.ts [hunk 5]
```diff
@@ -79,3 +99,58 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
...
+    const baseImageName = getBaseImageName(dockerFileContent);
+    const layers = dockerCommandUtils.getLayers(connection, imageUri);
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: Have separate block for getting pipeline variables and separate block for getting layers/base image name/imageUri

File: Tasks/DockerV2/dockerpush.ts [hunk 6]
```diff
@@ -62,12 +66,28 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
...
+
     // push all tags
     let output = "";
-    let promise = pushMultipleImages(connection, imageNames, tags, commandArguments, (commandOut) => output += commandOut);
-    
+    let outputImageName = "";
+    let digest = "";
+    let promise = pushMultipleImages(connection, imageNames, tags, commandArguments, (image, commandOutput) => {
+        output += commandOutput;
+        outputImageName = image;
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: What about the other images?

File: Tasks/DockerV2/dockerpush.ts [hunk 7]
```diff
@@ -79,3 +99,58 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
...
+        return response;
+    } catch (error) {
+        tl.debug('Unable to push to Image Details Artifact Store, Error: ' + error);
+    }
+
+    return Promise.resolve();
+}
+
+function extractDigestFromOutput(output: string): string {
+    const matchPatternForDigest = new RegExp(/sha256\:([\w]+)/);
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: can you please add the expected output from where you are fetching the sha?

File: Tasks/DockerV2/dockerpush.ts (hunk 1 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: This method uses sendRequest method from utilities-common/restutilities. If I move this to docker-common/pipelineUtils, I get the build errors, saying that these two util directories cannot have dependency on each other.
So I am placing this code in dockertask itself

File: Tasks/DockerV2/dockerpush.ts (hunk 2 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: Added the access token from 'SYSTEMVSSCONNECTION'

File: Tasks/DockerV2/dockerpush.ts (hunk 3 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: Refactored the code

File: Tasks/DockerV2/dockerpush.ts (hunk 4 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: Added another function getBaseImageNameFromDockerFileContent in containerImageUtils which reads the docker file and calls getBaseImageName

File: Tasks/DockerV2/dockerpush.ts (hunk 5 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: Done

File: Tasks/DockerV2/dockerpush.ts (hunk 6 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: I have moved pushImageDetails code inside onCommandOut which is called individually for each of the images inside a for loop.
Have verified for multiple images

**Comment by bansalaseem** [comment]
please add a good description of your change

File: Tasks/Common/docker-common/containerimageutils.ts [hunk 8]
```diff
@@ -26,6 +27,11 @@ export function generateValidImageName(imageName: string): string {
     return imageName;
 }
 
+export function getBaseImageNameFromDockerFileContent(dockerFilePath: string): string {
```
Comment (bansalaseem) [inline Tasks/Common/docker-common/containerimageutils.ts]: rename the method to getBaseImageNameFromDockerFile

File: Tasks/DockerV2/dockerpush.ts [hunk 9]
```diff
@@ -79,3 +102,78 @@ export function run(connection: ContainerConnection, outputUpdate: (data: string
...
+            "buildDefinitionName": buildDefinitionName,
+            "buildDefinitionId": buildDefinitionId
+        }
+    );
+
+    return sendRequestToImageStore(requestBody, requestUrl);
+}
+
+function extractDigestFromOutput(dockerPushCommandOutput: string): string {
+    const matchPatternForDigest = new RegExp(/sha256\:([\w]+)/);
```
Comment (bansalaseem) [inline Tasks/DockerV2/dockerpush.ts]: Add a comment talking about example output

File: Tasks/DockerV2/dockerpush.ts (hunk 1 above)
Comment (bansalaseem) [reply Tasks/DockerV2/dockerpush.ts]: In that case only move the part where you construct the requestBody

File: Tasks/Common/docker-common/containerimageutils.ts (hunk 8 above)
Comment (dabasnidhi) [reply Tasks/Common/docker-common/containerimageutils.ts]: done

File: Tasks/DockerV2/dockerpush.ts (hunk 9 above)
Comment (dabasnidhi) [reply Tasks/DockerV2/dockerpush.ts]: done

**Pull request merged**